### ISSUE TITLE
Need support for non-CPU accelerators in docker-compose 

### ISSUE THREAD

**Issue by vk1z** [compose-spec/compose-spec#74]
**What is the problem you're trying to solve**
The docker-compose specification does not support mixing GPU and CPU containers on the same machine. The runtime: tag that allowed us to do this was removed in 3.x. However, I believe that this is reflective of a larger problem that I'd like to see addressed. Devices will have multiple "accelerators" in addition to CPUs such as Nvidia GPU, AMD GPU, ARM GPU (pi), Google TPU, Intel TPU,... etc etc. Consequently, we need a clear way to request that a container gain access to a particular accelerator and that the orchestrator respects this when scheduling the container. Ideally orchestrators like Kube should also be able to play here.

**Describe the solution you'd like**
A run time constraint that describes the kind of accelerator needed will be ideal and the plumbing needed for the container to map the accelerator specified should be set up when the orchestrator schedules the container onto the machine hosting the specified accelerator. 

**Additional context**
Please see Github Issue: [Nvidia docker-compose](https://github.com/docker/compose/issues/6691)

**Comment by vk1z** [comment]
@chris-crone : Could you please let me know what else is needed? Getting an idea of timelines for this would be useful as well. It seems that we need to describe a constraint against the "accelerators" in the runtime environment that the container wants to have and for that one would need a uniform naming scheme for the kind of accelerators (nvidia-gpu,amd-gpu,google-tpu,intel-vpu,,,,) that were acceptable so that the right devices could be loaded.

**Comment by chris-crone** [comment]
Thanks for putting this up @vk1z!

>  Could you please let me know what else is needed?

The next step is to work with the community on finding a solution for GPU/TPU support (and maybe other devices) in the spec. We should sketch out the requirements and possibly some high level solutions here then create a pull request against the spec that adds the functionality.

As you pointed out, something like `runtime` doesn't solve the problem on different container platforms. There are broadly two approaches that I see:
1. We allow specifically attaching specific GPUs to a service (like we do for volumes, etc. in Compose)
1. We allow saying that a service requires a GPU to run and allow the platform to schedule the workload appropriately

(1) maps to something like what we have for volumes and networks, while (2) maps more to placement constraints. It's possible that we might want both in the spec for different use cases. Note that (1) is unlikely to map onto different platforms well.

Anecdotally, `docker run` [follows (1)](https://docs.docker.com/config/containers/resource_constraints/#gpu) and it appears that [AWS ECS GPU support](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-gpu.html) uses something like [Compose's placement constraints (2)](https://docs.docker.com/compose/compose-file/#placement).

@ndeloof have you done any thinking on this?

**Comment by vk1z** [comment]
@chris-crone : At least for kubernetes, google cloud seems to use a constraint to allocate [tpus or gpus](https://cloud.google.com/kubernetes-engine/docs/concepts/tpus). I like the explicit name prefix approach that google uses unlike the aws version that implicitly seems to assume that a gpu is nvidia.

**Comment by vk1z** [comment]
It seems to me that (1) and (2) address slightly different scenarios. (1) is for the case for manual (by the user) resource management. For example, the case where in docker we use --gpus and allocate gpus by hand to a container. (2) is as in the case where we have a resource manager in charge of scheduling containers who needs to be informed of user intent as in the case AWS ECS tasks or the google kubernetes link above. In the second case, we need some syntax to inform the resource manager of what accelerator and how many "units" need to be allocated given a group of machines controlled by said resource manager. It seems to me that the <manufacturer_name>/<accelerator_name> syntax has enough flexibility to accomodate this, in contrast to a more hard wired system as is present in docker CLI with --gpus or AWS ECS.  @chris-crone : How can we make faster progress on this? Should we get together on IRC and hash this out? Something else?

**Comment by chris-crone** [comment]
I've added this to the [agenda for the next meeting](https://hackmd.io/@ccrone/BylZwAROU). It could also be discussed on #compose-spec on the [CNCF Slack](http://slack.cncf.io/).

The process for getting this into the spec is defined [here](https://github.com/compose-spec/compose-spec/blob/master/CONTRIBUTING.md). Basically we need to get a proposal agreed and then merged into the spec.

Those making Compose tooling then need to implement that part of the spec.

**Comment by justincormack** [comment]
I think it might be useful to give the Docker CLI a GPU resource manager too. As GPUs become more partitionable this will probably become more important. It doesn't have to do much, just choose the first free GPU of the right type.

**Comment by EricHripko** [comment]
We've had a conversation about this with @justincormack at the recent Community Meeting and landed roughly on the conclusion that, until there's some standardisation in this space, it'd be difficult to add this to the Compose specification (since it'd end up as a collection of vendor-specific extensions). Should we close this for now and revisit at some point in the future?

**Comment by vk1z** [comment]
@EricHripko What kind of standardization is required? The problem is that unless this is supported, either people have to use ancient versions of docker-compose or script'ey hacks to get accelerators to work if they are not in a commercial cloud. I agree that it isn't perfect but supporting vendor namespaces for these seems like a reasonable first approach as is taken in GCP (not AWS).

**Comment by EricHripko** [comment]
See some notes [here](https://github.com/compose-spec/community/blob/master/meeting-notes/community/2020-05-21.md) - I'd summon Justin to elaborate too.

Could you explain what you mean by ancient versions of Compose?

**Comment by vk1z** [comment]
See this [link ](https://github.com/docker/compose/issues/6691#issuecomment-525406973). Since support for the nvidia runtime flag was removed thereafter, some folks who prefer to use compose and want access to nvidia gpu are still using 2.3. After all, we all have hard FPS requirements in this business ;-). 

I am NOT suggesting that the runtime flag that locks to Nvidia GPU is the right one. I also think that over time CPUs are likely to win anyway, but we still need something in the mean time to be able to use compose and satisfy our customers. I'm not saying that it is an easy problem to solve, but kicking the can down the road doesn't seem to be the right thing to do. The Google approach seems like the least worst of supporting such a vendor-specific-plugin model, which seems more or less a vendor specific namespace which names a vendor specific resource to load. Then at least, the problem isn't entirely a docker-compose problem. 

It is very likely that I'm missing something since my knowledge or understanding of the ramifications on docker compose may be lacking, but we genuinely need a way to run our business in the time between when now and when cpus do everything.

**Comment by EricHripko** [comment]
[runtime](https://github.com/compose-spec/compose-spec/blob/master/spec.md#runtime) attribute was brought back to the spec as a part of [v2-v3 merger](https://github.com/compose-spec/compose-spec/pull/78) - you should be able to use it once the changes to the spec are reflecting in tooling. I cannot imagine it will be removed without an alternative. Do you think this addresses the immediate needs for `docker-compose` + GPUs usage?

**Comment by vk1z** [comment]
Thanks @EricHripko , I was not aware of that. It will likely address the case for people who are running nvidia gpus. However, you should be aware of the fact that for edge inference (very common) scenario, that google is pushing tpus and integrating this fairly deeply into tensorflow-lite. So I suspect that the problem isn't going away soon. The issue is unfortunately rather tied to different business priorities across different competing vendors - nvidia, google, amd, intel etc. As practitioners, we are held to lowering cost per fps and so that causes us to need support for whatever makes the best business sense. For example, I cannot imagine deploying nvidia gpu for inference for most customers in cost sensitive countries, and we can't do cloud based inference due to latency and bandwidth restrictions. Just trying to provide some context for the overall problem and the incipient trainwreck :-) that can lie behind the runtime: approach if not thought out fully.

**Comment by EricHripko** [comment]
The latest comment on the linked issue has an interesting proposal for capability-based device requests https://github.com/docker/compose/issues/6691#issuecomment-670700674. We should explore whether this would work well with other Compose implementations.

**Comment by EricHripko** [comment]
As per [Community Meeting on Aug 27](https://hackmd.io/itoC6vBPQP6_3fV8B1vETA#2020-08-27), I'll look into formalising a proposal around `DeviceRequests` as a way to (hopefully) fill this gap.

See below for my notes so far:
- Realised that Docker doesn't support GPU acceleration on Docker Desktop
- Installed Linux on my desktop and got [TensorFlow](https://www.tensorflow.org/install/docker#gpu_support) to run with `docker run --gpu` invocation
- Researched how this functionality works in the daemon today:
  - It appears that nVidia support is [hard-coded](https://github.com/docker/docker-ce/blob/d49c4ca453ce4d1161ce5fb2482be7538bf73e07/components/engine/daemon/nvidia_linux.go) in the daemon. Based on comments, this is planned to be moved out into a Device Plugin in the future
  - `DeviceRequest` has the [following salient parameters](https://github.com/docker/docker-ce/blob/179b30ac853ff2c33d2b12acd8f71a53d05820e8/components/engine/api/types/container/host_config.go#L272): `driver` and `capabilities`
  - Given the above, the only usable `driver` today appears to be `nvidia`
  - Most capabilities appear to be [driver-specific](https://github.com/nvidia/nvidia-container-runtime#supported-driver-capabilities)
  - Daemon does add 2 more somewhat [generic capabilities](https://github.com/docker/docker-ce/blob/179b30ac853ff2c33d2b12acd8f71a53d05820e8/components/engine/daemon/nvidia_linux.go#L37): `gpu` and `nvidia`

**Comment by ndeloof** [comment]
`DeviceRequest` (as a flavour of `GenericResource`) seems to be the right abstraction level, but indeed the sole sample I can find so far are about NVidia GPU support. Maybe we should just embrace this major use case with
`DeviceRequest.capabilities: gpu` and allow implementations to define their own `x-*` extension for corner-cases so that we get a better understanding on actual use and needs, then the spec can be improved further.

**Comment by EricHripko** [comment]
I agree on abstraction level and do like the idea of making use of `x-*` extension fields. The one issue I've noticed with `capabilities: gpu` is that developers today use it to express `give my container a CUDA-enabled device` when this is a more generic capability. Given that MS will soon bring [dev/dxg](https://devblogs.microsoft.com/directx/directx-heart-linux/) to WSL2 and that AMD might support containers in the future, I don't think standardising just `gpu` is sufficient.

Perhaps, we should add something like this to the spec:
- `gpu` capability to select any GPU accelerator
- Provide a way to select CUDA-enabled accelerator by either:
  - Introducing a new `nvidia-cuda` or `nvidia.cuda` capability (Docker daemon will need to change to support this); or
  - Relying on combination of `gpu` + `nvidia` capabilities (supported today)

What are your thoughts?

**Comment by EricHripko** [comment]
Did some further experiments with the Engine API. These broadly followed the following pattern:
- Create a container with a `DeviceRequest` using low-level API:
```
$ curl -XPOST \
   --unix-socket /var/run/docker.sock \
   --header "Content-Type: application/json" \
   --data '{"Image":"tensorflow/tensorflow:latest-gpu","Cmd":["python","-c","import tensorflow as tf; print(tf.reduce_sum(tf.random.normal([1000, 1000])))"],"HostConfig":{"DeviceRequests": ... }]}}' \
   http://localhost/containers/create
{"Id":"...","Warnings":[]}
```
- Start the container to make sure that TensorFlow is using the GPU:
```
$ docker start -a ...
2020-09-05 10:16:41.330448: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1402] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 6602 MB memory) -> physical GPU (device: 0, name: GeForce RTX 2070, pci bus id: 0000:09:00.0, compute capability: 7.5)
tf.Tensor(-71.24344, shape=(), dtype=float32)
```

### Workflows

As a result, I outlined the following mappings to user workflows:
- Request device(s) with generic `gpu` capability:
```json
{"Capabilities" : [ [ "gpu" ] ] }
```
- Request device(s) with driver-specific `compute` capability:
```json
{"Capabilities" : [ [ "compute" ] ] }
```
- Request nVidia device(s) with driver-specific `compute` capability:
```json
{"Driver" : "nvidia", "Capabilities" : [ [ "compute" ] ]}
```
- Request specific device with generic `gpu` capability:
```json
{"Capabilities" : [ [ "gpu" ] ], "DeviceIDs" : [ "GPU-... ] }
```

**Comment by EricHripko** [comment]
Based on above, it seems that `compute` is the most accurate way to request a CUDA-enabled accelerator. However, this is quite a driver-specific capability and (I think) we should prefix driver-specific capabilities with the driver name to avoid name clashes in the future. We could use:
- The label approach: `nvidia.compute`
- The image approach: `nvidia/compute`
- Use dashes: `nvidia-compute`
